File: src/lava/magma/runtime/_c_message_infrastructure/test/test_cyclonedds_to_ros.cc
// Copyright (C) 2022 Intel Corporation
// SPDX-License-Identifier: BSD-3-Clause
// See: https://spdx.org/licenses/

#include <message_infrastructure/csrc/channel/dds/cyclone_dds.h>
#include <message_infrastructure/csrc/core/channel_factory.h>
#include <message_infrastructure/csrc/channel/dds/dds_channel.h>

using namespace message_infrastructure;  // NOLINT

#define LOOP_NUM 100

int main() {
  auto dds_channel = GetChannelFactory()
    .GetDDSChannel("rt/dds_topic",
                   DDSTransportType::DDSUDPv4,
                   DDSBackendType::CycloneDDSBackend,
                   10);
  auto dds_send = dds_channel->GetSendPort();
  int loop = LOOP_NUM;

  dds_send->Start();
  MetaDataPtr metadata = std::make_shared<MetaData>();
  metadata->nd = 1;
  metadata->type = 7;
  metadata->elsize = 1;
  metadata->total_size = 1;
  metadata->dims[0] = 1;
  metadata->strides[0] = 1;
  metadata->mdata = reinterpret_cast<char*>(malloc(sizeof(char)));
  while (loop--) {
    *reinterpret_cast<char*>(metadata->mdata) = loop % 255;
    dds_send->Send(metadata);
    printf("DDS send : '%d'\n", loop);
    sleep(1);
  }
  dds_send->Join();
  return 0;
}

File: src/lava/magma/runtime/_c_message_infrastructure/examples/ros2/test_cyclonedds_to_ros.py
# Copyright (C) 2022 Intel Corporation
# SPDX-License-Identifier: BSD-3-Clause
# See: https://spdx.org/licenses/

import numpy as np
import unittest
from functools import partial
import time

from lava.magma.runtime.message_infrastructure.multiprocessing \
    import MultiProcessing

from lava.magma.runtime.message_infrastructure import (
    ChannelBackend,
    Channel,
    SendPort,
    RecvPort,
    ChannelQueueSize,
    GetDDSChannel,
    DDSTransportType,
    DDSBackendType
)


def prepare_data():
    return np.random.random_sample((2, 4))


def test_ddschannel():
    name = 'rt/dds_topic'

    dds_channel = GetDDSChannel(
        name,
        DDSTransportType.DDSUDPv4,
        DDSBackendType.CycloneDDSBackend,
        ChannelQueueSize)

    send_port = dds_channel.src_port
    send_port.start()
    for i in range(100):
        send_port.send(prepare_data())
        time.sleep(0.1)
    send_port.join()


if __name__ == "__main__":
    test_ddschannel()
